### ISSUE TITLE
Investigate delegating TypeScript training binary readers to NEAT-AI-core training_bin_stream

### ISSUE THREAD

**Issue by stservice** [stSoftwareAU/NEAT-TS#2418]
## Summary
Scope out whether the TypeScript per-record `seekSync` / `readSync` path in `src/architecture/training/` should delegate to the new `training_bin_stream` module in `NEAT-AI-core` (PRs #28, #29), which adds `for_each_read_chunk_with_mode`, tunable I/O (`NEAT_SCORER_IO_MODE`, `NEAT_SCORER_READ_BYTES`), and a race-fixed pipelined double-buffer backend.

## What Needs to Be Done
- Write a short design note (in the issue comments or a draft doc under `docs/`) answering:
  - Which TS files currently walk `.bin` training files? (Confirmed candidates: `src/architecture/training/TrainingEpoch.ts`, `TrainingSetup.ts`, `TrainingSamples.ts`, `src/architecture/DataSet.ts`, `src/architecture/KFoldSplitter.ts`, `src/architecture/CrossValidationTrainer.ts`, `src/discovery/HoldoutValidator.ts`, `src/discovery/EnhancedDiscoveryValidator.ts`, `src/architecture/ErrorGuidedStructuralEvolution/DataRecorder.ts`.)
  - Is the new core module accessible from the WASM bundle, or is it strictly native (used only by `rust_scorer`)? Prior analysis suggests native-only, which means TS cannot call it directly today.
  - If native-only, what is the shape of the FFI or scorer-side integration that would be required? Is the cost worth the payoff versus the current Deno sync reader?
- Capture a benchmark baseline in `bench/` for the current Deno reader over a representative `.bin` corpus (read time, throughput). Baseline only — no code change.
- Decide: proceed (create a follow-up implementation issue referencing this one) or close as "no action" with the benchmark record attached. Per project policy, no PR for a performance-oriented change without before/after numbers.

## Acceptance Criteria
- [ ] Design note posted on this issue covering accessibility, integration shape, and cost/benefit.
- [ ] Benchmark baseline for current TS reader recorded (throughput, wall time on a fixed corpus) and attached to the issue.
- [ ] Explicit "proceed" or "no action" decision recorded. If proceed, a follow-up implementation issue is created and linked. If no action, this issue is closed with `negative-result` rationale.
- [ ] No production TS code is changed as part of this exploratory issue.

## Dependencies
Depends on #2414

## Context
Part of #2413

Core additions in scope: `for_each_read_chunk`, `for_each_read_chunk_with_mode`, `training_read_tuning_from_env`, `io_backend_label`, and `TrainingReadMode::{PipelinedDoubleBuffer, SingleBufferSequential}`. The key open question is whether TS can realistically benefit — the core reader is native (scorer path), and the existing TS reader is simple Deno sync I/O. Treating this as exploratory avoids committing to a migration that may not be worthwhile.

**Comment by stservice** [comment]
Claimed by `stservice-1777200914803`

**Comment by stservice** [comment]
## Investigation outcome — no action

Baseline benchmark + design note landed in PR #2444 (artefacts only, no production TS changed).

### TS files that walk `.bin` training files (confirmed)
Only `src/architecture/training/TrainingEpoch.ts` performs **per-record random access** (`seekSync` + `readSync`); `TrainingSetup.ts`, `TrainingSamples.ts`, `DataSet.ts`, `KFoldSplitter.ts`, `CrossValidationTrainer.ts`, `HoldoutValidator.ts`, `EnhancedDiscoveryValidator.ts`, `DataRecorder.ts`, and `DataRecorderRecording.ts` are sequential streams or append-only writers.

### Is `training_bin_stream` accessible from the WASM bundle?
**No.** Verified via ripgrep of the pinned `wasm_activation/pkg/wasm_activation.{js,d.ts}` artefacts — no exports for `training_bin_stream`, `for_each_read_chunk`, `TrainingReadMode`, or `io_backend_label`. The module is gated to the native `rust_scorer` crate. WASM in Deno cannot perform synchronous file I/O on its own anyway, so even an in-WASM port would route every read back through a JS host call, defeating the double-buffer pipelining design.

### Shape of a hypothetical native integration
1. **Deno FFI bridge** — new dynamic library on the build matrix (dylib/so/dll), per-record callback marshalling across the FFI boundary or shared chunk buffers, plus parity-gate coverage and version pinning. Adds ~1–10 µs per FFI call; even a 100 ns ideal call would cost ~150 ms over a 1.5 M-record epoch — a regression vs current ~1.8 s.
2. **Scorer-side integration** — move the entire training loop into Rust. Out of scope for an I/O optimisation.

### Baseline benchmark (current Deno sync reader)
`bench/binaryFormat/TrainingEpochReader.ts` over a 1 500 000-record / 3.63 GiB fixture (648 obs + 2 out per record, 2 600 bytes/record), Apple Silicon, page cache warm by run 3:

```text
Pattern                      records         time  throughput
sequential seekSync run 1       1500000 records     2.033 s      737 758 rec/s    1829.3 MiB/s
sequential seekSync run 2       1500000 records     1.812 s      828 001 rec/s    2053.1 MiB/s
sequential seekSync run 3       1500000 records     1.534 s      978 064 rec/s    2425.2 MiB/s
random sample seekSync 1         200000 records     0.295 s      678 785 rec/s    1683.1 MiB/s
random sample seekSync 2         200000 records     0.298 s      672 067 rec/s    1666.4 MiB/s
random sample seekSync 3         200000 records     0.313 s      638 750 rec/s    1583.8 MiB/s

Mean wall-clock time:
  sequential pass over 1500000 records : 1792.8 ms
  random sample of 200000 records       : 301.8 ms
```

Sequential reads land at **~2 GiB/s** (within the same order of magnitude as raw NVMe sequential); random-sample reads at **~1.7 GiB/s**. Per-epoch I/O is not the bottleneck — activation, backprop, and discovery dominate by orders of magnitude.

### Decision: no action

1. Core reader is native-only, not reachable from the WASM bundle.
2. A Deno FFI bridge would add per-call overhead with no I/O headroom to recover (the Deno reader is already at hardware bandwidth).
3. Project policy (Performance Task Workflow) forbids landing a performance-driven PR without measured improvement; no measurable improvement is possible here.

This is a negative result — no measurable improvement was found. Closing as `not planned` with the `negative-result` label. The benchmark and full design note are preserved at `bench/binaryFormat/TrainingEpochReader.ts` and `docs/issue-2418-training-bin-stream-investigation.md` (PR #2444) for future reference.

**Comment by stservice** [comment]
Closing as `not planned` — see PR #2444 and the comment above for the negative-result rationale and benchmark baseline.

**Comment by stservice** [comment]
Pull request https://github.com/stSoftwareAU/NEAT-AI/pull/2444 has been created to address this issue.
